test-tests/shared/src/test/scala/zio/test/LawsSpec.scala

```diff
@@ -0,0 +1,65 @@
+package zio.test
+
+import zio.ZIO
+import zio.test.Assertion._
+
+object LawsSpec extends ZIOBaseSpec {
+
+  implicit class BoolSyntax(l: Boolean) {
+    def ==>(r: Boolean): Boolean = r || !l
+  }
+
+  trait Equal[-A] {
+    def equal(a1: A, a2: A): Boolean
+  }
+
+  object Equal {
+    implicit val intEqual: Equal[Int]       = _ == _
+    implicit val stringEqual: Equal[String] = _ == _
+  }
+
+  implicit class EqualSyntax[A](self: A) {
+    def ===(that: A)(implicit eq: Equal[A]): Boolean =
+      eq.equal(self, that)
+  }
+
+  val reflexiveLaw = new Laws.Law[Equal]("reflexiveLaw") {
+    def apply[A](caps: Equal[A], a1: A, a2: A, a3: A) = {
+      implicit val eq: Equal[A] = caps
+      assert(a1 === a1)(isTrue)
+    }
+  }
+
+  val symmetryLaw = new Laws.Law[Equal]("symmetryLaw") {
+    def apply[A](caps: Equal[A], a1: A, a2: A, a3: A) = {
+      implicit val eq: Equal[A] = caps
+      assert((a1 === a2) ==> (a2 === a1))(isTrue)
+    }
+  }
+
+  val transitivityLaw = new Laws.Law[Equal]("transitivityLaw") {
+    def apply[A](caps: Equal[A], a1: A, a2: A, a3: A) = {
+      implicit val eq: Equal[A] = caps
+      assert(((a1 === a2) && (a2 === a3)) ==> (a1 === a3))(isTrue)
```

Comment:
If all the laws are going to be `assert(x)(isTrue)`, then it makes sense for them to return boolean values. It doesn't lead to helpful error messages (see also: ScalaCheck!), but printing out the "failure combination" will help a lot with that.

On the other hand, I feel like "downstream", it could make sense to introduce a ZIO Test DSL where `(a === a2)` actually generates a `TestResult` given `Equal[A]`. Or maybe use a different operator, e.g. Cats uses `<->`.

Reply:
Yes. This was just an example. In this case we can rewrite these to assert that the left hand side is equal to the right hand side so we get more helpful error messages. In general as you say we need an assertion that uses `Equal`. Question would be where that should live. I like the idea of calling it `===` but am worried that may create ambiguity or confusion with the `===` syntax for `Equal` so we may need to do something like `<->`.